apv5sdk-v15/apps/int6000-utils-1.1.6/int6k/int6kf.7
.TH int6kf 7 "Intellon Corporation" "int6000-utils-linux" "Intellon Linux Toolkit"
.SH NAME
int6kf - INT6300 Flash Utility for Linux
.SH SYNOPSIS
.BR int6kf
.RI [ options ] 
.BR -C 
.IR file 
.BR -N 
.IR file 
.BR -P 
.IR file
.SH DESCRIPTION
This program is deprecated. Use \fBint6kboot\fR when downloading firmware 3.3.4 or later on INT6300 devices or later.  
.PP
Download and start runtime firmware on a local Intellon powerline device that is in bootloader mode. A device stays in bootloader mode when it cannot load firmware from NVRAM. That can happen because there is no NVRAM or because the NVRAM is blank or corrupted. The bootloader only understands a subset of Intellon message types and cannot flash NVRAM. This program uses that subset to download runtime firmware and start execution. Optionally, it will proceed to flash NVRAM after the firmware has started. 
.PP
Booting a device is a two part operation. The first part downloads SDRAM configuration parameters then download runtime firmware and parameters into SDRAM and starts the firmware. The second part downloads runtime firmare and parameters (again) into scratch buffers and instructs the firmware to write them into to NVRAM. This program can perform the first part alone or the first then the second part but not the second part alone. The second part can be performed alone using program \fBint6k\fR.
.PP
This program is part of the Intellon Linux Toolkit. See the \fBilt\fR man page for compilation and installation instructions.
.SH OPTIONS
.TP
.RB - a
Wait until device sends a VS_HOST_ACTION.IND before downloading SDRAM configuration, firmware and PIB.
.TP 
-\fBC\fI filename\fR
Read the named SDRAM configuration file and write it to the device. This option and argument are required. Any file having an invalid SDRAM configuration file format will be rejected. No assumptions are made based on filename and no filename conventions are enforced.
.TP
.RB - e
Redirects stderr messages to stdout. By convention status and error messages are printed on stderr while primary program output is printed on stdout. This option prints all output on stdout in cases where this is desired.
.TP
.RB - F [ F ]
Write previously downloaded MAC and PIB to NVRAM using VS_MOD_NVM. Adding a second \fBF\fR here or another \fB-F\fR anywhere on the command line will \fBforce-flash\fR a blank or corrupted NVRAM. Firmware loaded from NVRAM will treat force-flash as an error. This option can be used to create factory settings but cannot be used to change them once created. Subsequent use creates and updates operational settings that can be erased using a factory reset. This option is executed after all others on the command line, except for the \fB-R\fR option.
.TP
-\fB\i \fIinterface\fR
Select the host Ethernet interface. All requests are sent via this interface and only reponses received via this interface are recognized. The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface. This option then takes precedence over either default.
.TP 
-\fBN \fIfilename\fR
Read the named NVM file and write it to the device. This option and argument are required. Any file having an invalid NVM file format will be rejected. No assumptions are made based on filename and no filename conventions are enforced.
.TP
-\fBP \fIfilename\fR
Read the named PIB file and write it to the device. This option and argument are required. Any file having an invalid PIB file format will be rejected. No assumptions are made based on filename and no filename conventions are enforced.
.TP
.RB - q
Suppresses status messages on stderr. 
.TP
.RB - v
Prints additional information on stdout. In particular, this option dumps outgoing Ethernet packets on stdout.
.TP
.RB - ? ,-- help
Print program help summary on stdout. This option takes precedence over other options on the command line. 
.TP
.RB - ! ,-- version
Print program version information on stdout. This option takes precedence over other options on the command line. Use this option when sending screen dumps to Intellon Technical Support so that they know exactly which version of the Linux Toolkit you are using.
.SH ARGUMENTS
None.
.SH REFERENCES
See the Intellon HomePlug AV Firmware Technical Reference Manual for more information.
.SH EXAMPLES
The \fB-C\fR, \fB-N\fR and \fB-P\fR options are always required. Their order is not important. Notice that no MAC address is needed because devices use address 00:B0:52:00:00:01, the program default address, when in bootloader mode.
.PP
   int6kf -C test.cfg -N test.nvm -P test.pib
   > eth0 00:B0:52:00:00:01 Write SDRAM Configuration
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Write to device memory
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Write to device memory
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Start Firmware
   > eth0 00:B0:52:00:00:01 Started.
.PP
Downloads SDRAM configuration file \fBtest.cfg\fR with VS_SET_SDRAM then firmware from file \fBtest.nvm\fR with VS_WR_MEM then PIB from file \fBtest.pib\fR then starts firmware execution with VS_ST_MAC. This operation is common on flash-less systems.
.PP
   int6kf -C test.cfg -N test.nvm -P test.pib -F
   > eth0 00:B0:52:00:00:01 Write SDRAM Configuration
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Write to device memory
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Write to device memory
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Start Firmware
   > eth0 00:B0:52:00:00:01 Started.
   > eth0 00:B0:52:00:00:01 Write MAC as module
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Write PIB as module
   > eth0 00:B0:52:00:00:01 Written.
   > eth0 00:B0:52:00:00:01 Flash Device
   > eth0 00:B0:52:00:00:01 Flashed.
.PP
Perform the same actions as before with some extra steps. After starting firmware execution, the same firmware and PIB are downloaded again, using VS_WR_MOD instead of VS_WR_MEM, and written to NVRAM using VS_MOD_NVM with force. The force option is only recognized by firmware that has been started by the bootloader and will fail when no flash memory is present.
.SH DISCLAIMER
Intellon HomePlug AV Vendor Specific Management Message Entry structure and content is proprietary to Intellon Corporation, Ocala FL USA. Consequently, public information may not available. Intellon Corporation reserves the right to modify message structure and content in future firmware releases without any obligation to notify or compensate users of this program.
.SH SEE ALSO
.BR ilt ( 7 ),
.BR int6k ( 7 ),
.BR int6kid ( 7 ),
.BR int6krate ( 7 ),
.BR int6kwait ( 7 )
.SH CREDITS
 Charles Maier <charles.maier@intellon.com>


apv5sdk-v15/apps/int6000-utils-1.1.6/int6k/int6kf.c
/*====================================================================*
 *   
 *   Copyright (c) 2010, Atheros Communications Inc.
 *   
 *   Permission to use, copy, modify, and/or distribute this software 
 *   for any purpose with or without fee is hereby granted, provided 
 *   that the above copyright notice and this permission notice appear 
 *   in all copies.
 *   
 *   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL 
 *   WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL  
 *   THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR 
 *   CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM 
 *   LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, 
 *   NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 *   CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *   
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   int6kf.c - Intellon INT6300 Device Flash Utility;
 *
 *   this program sends and receives raw ethernet frames and so needs
 *   root priviledges; if you install it using "chmod 555" and "chown 
 *   root:root" then you must login as root to run it; otherwise, you 
 *   can install it using "chmod 4555" and "chown root:root" so that 
 *   anyone can run it; the program will refuse to run until you get
 *   things right;
 *   
 *   This software and documentation is the property of Intellon 
 *   Corporation, Ocala, Florida. It is provided 'as is' without 
 *   expressed or implied warranty of any kind to anyone for any 
 *   reason. Intellon assumes no responsibility or liability for 
 *   errors or omissions in the software or documentation and 
 *   reserves the right to make changes without notification. 
 *   
 *   Intellon customers may modify and distribute the software 
 *   without obligation to Intellon. Since use of this software 
 *   is optional, users shall bear sole responsibility and 
 *   liability for any consequences of it's use. 
 *   
 *.  Intellon INT6x00 Linux Toolkit for HomePlug AV; 
 *:  Published 2007 by Intellon Corp. ALL RIGHTS RESERVED;
 *;  For demonstration; Not for production use;
 *
 *   Contributor(s):
 *      Charles Maier <charles.maier@intellon.com>
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../int6k/int6k.h"
#include "../ram/nvram.h"
#include "../ram/sdram.h"
#include "../nvm/nvm.h"
#include "../pib/pib.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../int6k/Confirm.c"
#include "../int6k/chipset.c"
#include "../int6k/Display.c"
#include "../int6k/Failure.c"
#include "../int6k/FlashDevice.c"
#include "../int6k/FlashNVM.c"
#include "../int6k/ReadMME.c"
#include "../int6k/Request.c"
#include "../int6k/SendMME.c"
#include "../int6k/StartDevice.c"
#include "../int6k/StartFirmware.c"
#include "../int6k/UpgradeDevice.c"
#include "../int6k/WriteCFG.c"
#include "../int6k/WriteMEM.c"
#include "../int6k/WriteNVM.c"
#include "../int6k/WritePIB.c"
#include "../int6k/WaitForReset.c"
#include "../int6k/WaitForStart.c"
#include "../int6k/EthernetHeader.c"
#include "../int6k/IntellonHeader.c"
#include "../int6k/UnwantedPacket.c"
#endif

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/checkfilename.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/todigit.c"
#include "../tools/hexdump.c"
#include "../tools/checksum_32.c"
#include "../tools/fdchecksum_32.c"
#include "../tools/error.c"
#include "../tools/alert.c"
#include "../tools/strfbits.c"
#include "../tools/filepart.c"
#endif

#ifndef MAKEFILE
#include "../ram/sdramfile.c"
#include "../ram/sdrampeek.c"
#endif

#ifndef MAKEFILE
#include "../nvm/nvmfile.c"
#include "../nvm/nvmpeek.c"
#endif

#ifndef MAKEFILE
#include "../pib/pibcodes.c"
#include "../pib/pibfile.c"
#include "../pib/pibpeek.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

#ifndef MAKEFILE
#include "../mme/MMECode.c"
#endif

#ifndef MAKEFILE
#include "../crypt/keys.c"
#endif

/*====================================================================*
 *   
 *   int main (int argc, const char * argv[]);
 *   
 *   parse command line, populate int6k structure and perform selected 
 *   operations; show help summary when asked; see getoptv and putoptv
 *   to understand command line parsing and help summary display; see
 *   int6k.h for the definition of struct int6k; 
 *
 *   the default interface is eth1 because most people use eth0 as 
 *   their principle network connection; you can specify another 
 *   interface with -i or define environment string PLC to make
 *   that the default interface and save typing;
 *   
 *.  Intellon INT6x00 Linux Toolkit for HomePlug AV; 
 *:  Published 2007 by Intellon Corp. ALL RIGHTS RESERVED;
 *;  For demonstration; Not for production use;
 *
 *--------------------------------------------------------------------*/

int main (int argc, const char * argv []) 

{
	extern struct channel channel;
	extern const char * chipset [CHIPSETS+1];
	static const char *optv [] = 
	{
		"C:i:eFN:p:P:qt:vx",
		"-C file -P file -N file",
		"Intellon INT6300 Flash Utility",
		"C f\twrite CFG file to device using VS_SET_SDRAM",
		"e\tredirect stderr messages to stdout",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface number [2]",

#else

		"i s\thost interface name [" CHANNEL_ETHDEVICE "]",

#endif

		"F[F]\tflash [force] NVRAM after firmware start using VS_MOD_NVM",
		"N f\twrite NVM file to device using VS_WR_MEM",
		"P f\twrite PIB file to device using VS_WR_MEM",
		"q\tquiet mode",

#if defined (WINPCAP) || defined (LIBPCAP)

		"t n\tread capture time is (n) milliseconds [50]",

#else

		"t n\tread timeout is (n) milliseconds [50]",

#endif

		"v\tverbose mode",
		"x\texit on error",
		(const char *) (0) 
	};

#include "../int6k/int6k-struct.c"

	char string [INT6K_VERSTRING];
	signed c;
	if (getenv (PLCDEVICE)) 
	{
#if defined (WINPCAP) || defined (LIBPCAP)
		channel.id = atoi (getenv (PLCDEVICE));
#else
		channel.name = strdup (getenv (PLCDEVICE));
#endif
	}
	optind = 1;
	opterr = 1;
	while ((c = getoptv (argc, argv, optv)) != -1) 
	{
		switch ((char) (c)) 
		{
		case 'C':
			if (!checkfilename (optarg)) 
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((int6k.CFG.file = open (optarg, O_BINARY|O_RDONLY)) == -1) 
			{
				error (1, errno, "%s", optarg);
			}
			if (sdramfile (int6k.CFG.file, optarg, int6k.flags)) 
			{
				error (1, ECANCELED, "CFG file %s is corrupt", optarg);
			}
			_setbits (int6k.flags, INT6K_SDRAM_CONFIG);
			int6k.CFG.name = optarg;
			break;
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.id = atoi (optarg);

#else

			channel.name = optarg;

#endif

			break;
		case 'F':
			_setbits (int6k.modulecode, VS_MODULE_ALL);
			if (_anyset (int6k.flags, INT6K_FLASH_DEVICE)) 
			{
				_setbits (int6k.modulecode, VS_MODULE_FORCE);
			}
			_setbits (int6k.flags, INT6K_FLASH_DEVICE);
			break;
		case 'N':
			if (!checkfilename (optarg)) 
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((int6k.NVM.file = open (optarg, O_BINARY|O_RDONLY)) == -1) 
			{
				error (1, errno, "%s", optarg);
			}
			int6k.NVM.name = optarg;
			if (nvmfile (&int6k.NVM)) 
			{
				error (1, errno, "Bad NVM file: %s", int6k.NVM.name);
			}
			_setbits (int6k.flags, INT6K_WRITE_MAC);
			break;
		case 'P':
			if (!checkfilename (optarg)) 
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((int6k.PIB.file = open (optarg, O_BINARY|O_RDONLY)) == -1) 
			{
				error (1, errno, "%s", optarg);
			}
			int6k.PIB.name = optarg;
			if (pibfile (&int6k.PIB)) 
			{
				error (1, errno, "Bad PIB file: %s", int6k.PIB.name);
			}
			_setbits (int6k.flags, INT6K_WRITE_PIB);
			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			_setbits (int6k.flags, INT6K_SILENCE);
			break;
		case 't':
			channel.timeout = (unsigned)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (int6k.flags, INT6K_VERBOSE);
			break;
		case 'x':
			_setbits (int6k.flags, INT6K_BAILOUT);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc) 
	{
		error (1, ECANCELED, "Too many arguments");
	}
	if (int6k.CFG.file == -1) 
	{
		error (1, ECANCELED, "No CFG file specified");
	}
	if (int6k.PIB.file == -1) 
	{
		error (1, ECANCELED, "No PIB file specified");
	}
	if (int6k.NVM.file == -1) 
	{
		error (1, ECANCELED, "No NVM file specified");
	}

#ifndef WIN32

/*
 *      raw packets require root priviledges on linux; one does not have to be
 *      root when this program is installed; setuid using 'chown root:root' and 
 *      'chmod 4555';
 */

	if (geteuid ()) 
	{
		error (1, EPERM, INT6K_NEEDROOT);
	}

#endif

	openchannel (&channel);
	if (!(int6k.message = malloc (sizeof (struct message)))) 
	{
		error (1, errno, INT6K_ERR_MEMORY);
	}
	if (WaitForStart (&int6k, string, sizeof (string))) 
	{
		Failure (&int6k, "Device must be connected");
		return (-1);
	}

#if 1

	if (int6k.hardwareID > INT6300A0) 
	{
		Failure (&int6k, "Device must be %s or earlier; try using int6kboot.", chipset [INT6300A0]);
		return (-1);
	}

#endif

	if (strcmp (string, "BootLoader")) 
	{
		Failure (&int6k, "Bootloader must be running");
		return (-1);
	}
	if (!StartDevice (&int6k)) 
	{
		if (_anyset (int6k.flags, INT6K_FLASH_DEVICE)) 
		{
			UpgradeDevice (&int6k);
		}
	}
	free (int6k.message);
	closechannel (&channel);
	exit (0);
}

